daemon/algod/api/algod.oas2.json

```diff
@@ -2429,6 +2429,89 @@
           }
         }
       }
+    },
+    "/v2/devmode/blocks/offset": {
+      "get": {
+        "description": "Gets the current timestamp offset.",
+        "tags": [
+          "public",
+          "nonparticipating"
+        ],
+        "produces": [
+          "application/json"
+        ],
+        "schemes": [
+          "http"
+        ],
+        "summary": "Returns the timestamp offset. Timestamp offsets can only be set in dev mode.",
+        "operationId": "GetBlockTimeStampOffset",
+        "responses": {
+          "200": {
+            "description": "OK",
+            "$ref": "#/responses/GetBlockTimeStampOffsetResponse"
+          },
+          "400": {
+            "description": "TimeStamp offset not set.",
+            "schema": {
+              "$ref": "#/definitions/ErrorResponse"
+            }
+          },
+          "default": {
+            "description": "Unknown Error"
+          }
+        }
+      }
+    },
+    "/v2/devmode/blocks/offset/{offset}": {
+      "post": {
+        "description": "Sets the timestamp offset (seconds) for blocks in devmode. Setting the offset to 0 will unset this value and try to use the real clock for the timestamp.",
+        "tags": [
+          "public",
+          "nonparticipating"
+        ],
+        "schemes": [
+          "http"
+        ],
+        "summary": "Given a timestamp offset in seconds, adds the offset to every subsequent block header's timestamp.",
+        "operationId": "SetBlockTimeStampOffset",
+        "parameters": [
+          {
+            "type": "integer",
+            "description": "The timestamp offset for blocks in devmode.",
```

Comment:
Maybe this is a minor point, but I would guess that devs want to set the time of the next block  _precisely_, because the intended use case isa unit test like "Set the timestamp, then call my contract that computes based on its value, and check the result"

By setting the _offset_, the caller doesn't know if maybe the next block will be one second later, based on bad luck.

I'd recommend that the external API is to set an exact time. That time is used in the next block, and only at that point is the offset computed to use from there on.

Reply:
We took a mini community poll here regarding how the API should behave: https://github.com/algorand/go-algorand/issues/3192#issuecomment-1501901340

Your point came up in thread, but there seemed to be agreement that the deltas can achieve this in a roundabout way. Also, I didn't want to allow negative offsets or rewinding timestamps for now - but if there is a compelling use case, maybe we should allow completely arbitrary timestamps in the block header?

Reply:
I think Fergal gave up too easily. Unless I'm wrong about implementation, the proposed use of offset does not meet the use case of a deterministic block.

Reply:
Also left a comment in the thread - happy to do what the community wants/needs at the end of that discussion

Reply:
Ok, I think I understand and agree this is fine now, sorry for all the chatter.